### PR TITLE
feat(core-flows, fulfillment): validate shipping options to stock locations

### PR DESCRIPTION
**Pull request by riqwan**
what:

- validate shipping options to stock locations

RESOLVES CC-232

### CODE REVIEW

File: integration-tests/http/__tests__/shipping-option/admin/shipping-option.spec.ts [hunk 1]
```diff
@@ -331,6 +406,73 @@ medusaIntegrationTestRunner({
             })
           )
         })
+
+        it.only("should throw an error when provider does not belong to service location", async () => {
```
Comment (sradevski) [inline integration-tests/http/__tests__/shipping-option/admin/shipping-option.spec.ts]:
```suggestion
        it("should throw an error when provider does not belong to service location", async () => {
```

File: integration-tests/http/__tests__/shipping-option/admin/shipping-option.spec.ts [hunk 2]
```diff
@@ -359,6 +501,22 @@ medusaIntegrationTestRunner({
             rules: [shippingOptionRule],
           }
 
+          const remoteLink = appContainer.resolve(
+            ContainerRegistrationKeys.REMOTE_LINK
+          )
+
+          // TODO: move this to an endpoint when available
+          await remoteLink.create([
```
Comment (sradevski) [inline integration-tests/http/__tests__/shipping-option/admin/shipping-option.spec.ts:509]: suggestion: Maybe create a helper func so we have one place to update when the endpoint is available

File: packages/core/core-flows/src/fulfillment/steps/validate-fulfillment-providers.ts [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+  "validate-fulfillment-providers-step"
+export const validateFulfillmentProvidersStep = createStep(
+  validateFulfillmentProvidersStepId,
+  async (input: FulfillmentProviderValidationInput[], { container }) => {
+    const dataToValidate: {
+      service_zone_id: string
+      provider_id: string
+    }[] = []
+    const shippingOptionIds = input.map((d) => d.id).filter(Boolean) as string[]
+    const fulfillmentService = container.resolve(
```
Comment (sradevski) [inline packages/core/core-flows/src/fulfillment/steps/validate-fulfillment-providers.ts:25]: Super NIT: maybe move `shippingOptionsIds` below `fulfillmentService`, it's a bit easier to read 😄. Maybe even passing the map to `id: input.map...` would be fine as well, WDYT?

File: packages/core/core-flows/src/fulfillment/steps/validate-fulfillment-providers.ts [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+    )
+    const shippingOptions = await fulfillmentService.listShippingOptions({
+      id: shippingOptionIds,
+    })
+
+    const shippingOptionsMap = new Map<string, ShippingOptionDTO>(
+      shippingOptions.map((so) => [so.id, so])
+    )
+
+    const invalidProviders: string[] = []
```
Comment (sradevski) [inline packages/core/core-flows/src/fulfillment/steps/validate-fulfillment-providers.ts]: Similarly small - move the declaration of `invalidProviders` further down where it is used

File: packages/core/core-flows/src/fulfillment/steps/validate-fulfillment-providers.ts [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+          provider_id:
+            data.provider_id! || existingShippingOption?.provider_id!,
+        })
+
+        continue
+      }
+
+      if (data.service_zone_id && data.provider_id) {
+        dataToValidate.push({
+          service_zone_id: data.service_zone_id!,
```
Comment (sradevski) [inline packages/core/core-flows/src/fulfillment/steps/validate-fulfillment-providers.ts]: I would have imagined the `!` to not be necessary here?

**Review by sradevski — APPROVED** [review]
LGTM! Welcome back btw 😄

File: integration-tests/http/__tests__/shipping-option/admin/shipping-option.spec.ts (hunk 2 above)
Comment (riqwan) [reply integration-tests/http/__tests__/shipping-option/admin/shipping-option.spec.ts:509]: I've already replaced these with the endpoint in the new branch i'm working on 🤞🏻

**Review by olivermrbl — APPROVED** [review]
LGTM, nice work. 

A minor thing on UI-side (for a follow-up PR): I believe we also need to filter out selectable providers in the UI when creating and updating shipping options. We should only show the providers that are associated with the location for which you are currently manage shipping options.

**Comment by riqwan** [comment]
@olivermrbl I've created a ticket for it in the project 🤞🏻

**Comment by riqwan** [comment]
@olivermrbl this PR is essentially a breaking change for shipping options, do we have a way to communicate this?

**Pull request merged**